---
title: "Assignment of Proteins to Subcellular Locations Using protlocassign"
author: "Dirk Moore"
date: "`r Sys.Date()`"
#output: rmarkdown::html_vignette
output: word_document

#output: 
#  word_document:
#    reference_docx: word-styles-reference-03.docx
#output: pdf_document
#fontsize: 10pt

#vignette: >
#  %\VignetteIndexEntry{protlocassign}
#  %\VignetteEngine{knitr::rmarkdown}
#  %\VignetteEncoding{UTF-8}
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```
Determining the locations of proteins in the cell is an important but complex problem. The major biochemical approach for this involves centrifugation-based methods to partially separate different organelles and other cellular compartments and then to determine the relative distribution of different proteins among the differential fractions. The location of proteins of interest are then inferred by comparing their distribution to those of known locations. This package takes data from centrifugation fractions for large numbers of proteins to determine their relative abundance among the fractions, and by extension to their relative abundance among subcelluar locations. 

This package implements a subcellular protein assignment procedure as "constrained proportional assignmet", or CPA (Jadot et al, 2016). The input data consists of a file with one column, geneName, of gene/protein identifiers, or of peptide identifiers, and multiple columns of the relative abundances of the proteins among the centrifugal fractions. The sum of the abundance levels is constrained to 1 but can be further transformed to improve the quality of the fit. Another key input to the program is a curated list of reference genes, and their subcellular locations. The method first determines the profiles of abundance levels of genes in a curated reference set. Then, for each gene, it finds either a profile matching a single compartment or a linear combination of compartment profiles that match the profile of that gene. The relative weights of the linear combination in principle reflect the relative abundance of a given protein among different compartments. The CPA method, unlike other previously described methods, can thus account for proteins that have multiple residences, and estimate the relative proportion among these residences. The method can also be applied to classify individual peptides, gaining potential insight into subcellular location of protein isoforms. There is also a tool to compute nearest neighbor distances of proteins, allowing grouping of proteins with similar distributions.



## Tutorial on gene assignment

We illustrate with an example. Tannous et al. (2020) presented abundance levels of proteins among six fractions from a differential centrifugation experiment (N, M, L1, L2, P, and S), and three fractions from a Nycodenz density gradient separation of the differential fraction L1 (Nyc1, Nyc2, and Nyc3). Eight subcellular compartments were considered: nucleus (Nuc), mitochondria (Mito), lysosomes (Lyso), peroxisomes (Perox), endoplasmic reticulum (ER) Golgi apparatus (Golgi), plasma membrane (PM), and cytosol (Cyto). The CPA method assigns each of a large number of genes to one or more of these compartments, based on profiles from a set of reference genes. 

## Installing the protlocassign package
Start by installing the devtools package from CRAN, by typing

```
install.packages("devtools")
```

Then install the protlocassign package from the github repository by typing

```
devtools::install_github("mooredf22/protlocassign0p1p1")
```

This will make the programs and datasets available.
Also, the `BB` library is required; it may be downloaded from CRAN by typing

```
install.packages("BB")
```
Once you have installed a package, you do not need to re-install it the next time you start up R.

These are the only packages you need to run the programs that work with data that are presented as mean levels for a protein/gene. In another vignette there is a description of R functions that work with data at the spectral level; additional packages will be needed to process spectral-level data and produce gene-level data.

## Working with data

To use this package you will need two data sets. One consists of a column of gene names followed by a series of columns of relative protein levels derived from a subcellular fractionation experiment. The other data set consists of a list of reference proteins in the first column and the name of the known subcellular compartment in the second column.

Consider for example the TMT MS2 data from Tannous et al. (2020). To get started, load the package and look at the dimensions of the data set and then the last few rows:

```{r, echo=TRUE}
library(protlocassign)
dim(geneProfileSummaryTMTms2)
```

Now use the `str` function to summarize the contents of `geneProfileSummaryTMTms2`:

```{r, echo=TRUE}
options(digits=2)
str(geneProfileSummaryTMTms2)
```

Each row represents a column, and presents the variable name, type (character or number) and the first few values of that variable. The first column is the gene name and the next eight columns represent the relative abundance level of the gene product. The last two columns, which are optional, give the number of spectra and the number of peptides (sequences) for each gene. (The spectra were averaged using a nested random effects model described in Jadot, 2016, to produce the given average for each gene.) 

Alternatively, you may also examine the first few rows by typing `head(geneProfileSummaryTMTms2)`.

The second data set we need has 39 reference genes and two columns, the first for the gene name and the second for their respective subcellular compartments. These are from Jadot et al (2016) and may be viewed as follows:
```{r, echo=TRUE}
dim(refLocProteinsJadot)
refLocProteinsJadot
```

The protlocassign package includes several other relative fraction data sets. In addition, you may read in your own data. Suppose for example that the data sets reside in a directory `"C:\temp\myproteindata"`. First set the working directory to point there:

```
setwd("C:\\temp\\myproteindata")
```

Note that in R each backslash character must be doubled to be interpreted correctly (since otherwise it will be incorrectly interpreted as an escape character). Alternatively, you may use a single foward slash:

```
setwd("C:/temp/myproteindata")
```

Then read in the two data sets (profile data and reference genes), which must be in
comma-separated format (`.csv`), with the first row containing column names:

```
geneProfileSummaryMyProfiles <- read.csv(file="geneProfileSummaryMyProfiles.csv",
                                         row.names=F
refLocProteinsMyRefs <- read.csv(file="refLocProteinsMyRefs.csv", row.names=F)
```

## Obtaining constrained proportionate assignments of gene products to compartments

In order to assign proteins proportionately to their respective compartments, we first must obtain profiles for the reference genes, use the function `cpaSetup`, which averages the spectra in each reference gene. This matrix has one row for each compartment and one column for each fraction.

```{r, echo=TRUE, eval=TRUE}
markerLocRuse <- cpaSetup(geneProfileSummary=geneProfileSummaryTMTms2, refLocProteins=refLocProteinsJadot, n.channels=9)
markerLocRuse

```

To view a particular compartment profile and its component genes, use `refProfilePlot`. To examine the available compartments, use the row names of markerLocR. 

```{r, echo=TRUE, eval=TRUE, fig.show='hold'}
rownames(markerLocRuse)
```

For example, to plot the profile for cytosol, use this:

```{r, echo=TRUE, eval=TRUE, fig.show='hold'}

refProfilePlot(refLoc="PM", refLocProteins=refLocProteinsJadot,
                     geneProfileSummary=geneProfileSummaryTMTms2,
                     markerLocR=markerLocRuse)
```

```{r, echo=TRUE, eval=TRUE, fig.show='hold'}
windows(height=10, width=7)
par(mfrow=c(3,2))
for (j in 1:5) {
  refProfilePlot(refLoc="PM", refLocProteins=refLocProteinsJadot,
                     geneProfileSummary=geneProfileSummaryTMTms2,
                     markerLocR=markerLocRuse, refGenePlot=j)
  }
```

To plot all of the reference profiles we may set up a plot window with four rows and two columns and then loop through the eight subcellular compartments. (Plots not shown here.)

```{r, echo=TRUE, eval=FALSE}
loc.list <- rownames(markerLocRuse)
n.loc <- length(loc.list)
windows(width=7, height=10)
par(mfrow=c(4,2))
for (i in 1:n.loc) {
  refProfilePlot(refLoc=loc.list[i], refLocProteins=refLocProteinsJadot,
                     geneProfileSummary=geneProfileSummaryTMTms2,
                     markerLocR=markerLocRuse)
  }
```

Now we can run the CPA routine, which assigns proteins to their respective compartments; this may take several minutes to complete. 

```{r, echo=TRUE, eval=TRUE, message=FALSE, warning=FALSE}
assignPropsUse <- proLocAll(geneProfileSummary=geneProfileSummaryTMTms2,
                            markerLocR=markerLocRuse, n.channels=9)

str(assignPropsUse)
```

The first column contains the gene names, and the next 8 columns contain the allocation proportions of each gene to the eight compartments. The last two columns are the number of peptides and spectra, carried over from the input `geneProfileSummary` profile data, which is not essential for the CPA analysis.

We can look at the profile of, for example, the gene "TLN1". We can first find the gene in the dataset, and also the row number of that gene:

```{r, echo=TRUE, eval=TRUE}
protIndex("TLN1")
#protPlotfun(protPlot=93, assignProbsOut=assignProbsOut)
```

This function also accepts partial matching of the first few letters of a gene. For example, we can find the indices of the genes starting with "TLN":

```{r, echo=TRUE, eval=TRUE}
protIndex("TLN")
```



Now we plot the results for gene number TLN1:

```{r, echo=TRUE, eval=TRUE, fig.width=7, fig.height=7}
#protPlotfun(protName="TLN1",
# geneProfileSummary=geneProfileSummaryTMTms2, 
#n.fractions=9, n.compartments=8,
# Nspectra=T, markerLocR=markerLocRuse, assignPropsMat=assignPropsUse)
```                        

The horizontal axis represents the nine fractions, which are N, M, L1, L2, P, S, Nyc.1, Nyc.2, and Nyc.3. In each of the eight plots, the red line is the average profile of the peptides, which are represented by the blue lines; thicker blue lines indicate larger numbers of spectra for a particular peptide. The dashed yellow-black lines show the expected profile for a protein entirely resident in the respective subcellular location. In this set of plots, we see that the CPA procedure assigns a 59 percent residence proportion to plasma membrane and 35 percent residence to cytosol. Visually, we see that the observed red profile closely matches a mixture of the expected yellow-blacklines. Also note that there is one peptide, plotted in solid blue, with a quite different profile from the other peptides. This is likely due to a peptide (specifically, MVAAAK) incorrectly assigned to this protein. This peptide was removed from the estimation of the average profile by an outlier rejection procedure. This outlier rejection procedure is discussed in another vignette.

## Nearest genes in a data set
We may find the genes with profiles nearest to a given gene using the function "nearestGenes". Distance is computed as the Euclidean distance between profiles. To use the function, we first create a distance matrix for the genes in a list of mean profiles, such as "geneProfileSummaryTMTms2". In this dataset, the profiles are in columns 2 through 10.

```{r, echo=TRUE, eval=TRUE}
distUse <- dist(geneProfileSummaryTMTms2[,2:10], method="euclidean")
```
Then select the gene names:

```{r, echo=TRUE, eval=TRUE}
genesUse <- geneProfileSummaryTMTms2[,1]
```
Finally, provide a gene name. Here, we first look for genes in our dataset that start with "TLN". We choose the gene "AADAC", and find the 10 nearest genes.
```{r, echo=TRUE, eval=TRUE}
protIndex("TLN")
nearestGenes("TLN1", n.nearest=10,  distGenes=distUse, geneNames=genesUse)
```

We now consider the Christofrou data, which has 20 channels (from two combined experiments). We remove GPT_MOUSE and CPSM_MOUSE (rows 2 and 18). Because of the large number of columns, the output text wraps around to display all of them.

```{r, echo=TRUE, eval=TRUE}
refLocProteinsJadotMouseMod <- refLocProteinsJadotMouse[-c(2,18),] 
# remove row 2 and 18  (GPT_MOUSE and CPSM_MOUSE)
markerLocRuse <- cpaSetup(geneProfileSummary=geneProfileSummaryChristoforou,
                       refLocProteins=refLocProteinsJadotMouse, n.channels=20)
markerLocRuse
```

For example, to plot the profile for cytosol, use this:

```{r, echo=TRUE, eval=TRUE, fig.show='hold'}

refProfilePlot(refLoc="Cyto", refLocProteins=refLocProteinsJadotMouse,
                     geneProfileSummary=geneProfileSummaryChristoforou,
                     markerLocR=markerLocRuse)
```

We can see that there is one protein labelled "Cyto" that appears different from the others, and may represent a data errror.



As we did earlier, we may plot the profiles of all of the compartments for which at least one marker protein was found. (Plots not shown here.)

```{r, echo=TRUE, eval=FALSE}
loc.list <- rownames(markerLocRuse)
n.loc <- length(loc.list)
windows(width=7, height=10)
par(mfrow=c(3,2))
for (i in 1:n.loc) {
  refProfilePlot(refLoc=loc.list[i], refLocProteins=refLocProteinsJadotMouse,
                     geneProfileSummary=geneProfileSummaryChristoforou,
                     markerLocR=markerLocRuse)
  }
```


Next we estimate the CPA proportions for this experiment.

```{r, echo=TRUE, eval=TRUE}
assignPropsUse <- proLocAll(geneProfileSummary=geneProfileSummaryChristoforou,
                                markerLocR=markerLocRuse, n.channels=20)

str(assignPropsUse)
```

Now we can plot the CPA estimates from the Christoforou data.

```{r, echo=TRUE, eval=TRUE}
#protPlotfun(protName="TLN1_MOUSE", geneProfileSummary=geneProfileSummaryChristoforou,
#n.fractions=20, n.compartments=6,
#            Nspectra=T, markerLocR=markerLocRuse, assignPropsMat=assignPropsUse)
```

This data set yields estimates of 73 percent plasma membrane and 22 percent cytosol, which are comparable to the rates from Tannous et al discussed above (59 percent and 35 percent, respectively).

## References

Christoforou A, Mulvey, CM, Breckels LM, Geladaki A, Hurrell T, Hayward PC, Naake T, Gatto L, Viner R, Arias AM, and Lilley KS (2016), "A draft map of the mouse pluripotent stem cell spatial proteome" Nature Communications 7, 8992. DOI: 10.1038/ncomms9992

Jadot M, Boonen M, Thirion J, Wang N, Xing J, Zhao C, Tannous A, Qian M, Zheng H, Everett JK, Moore DF, Sleat DE, Lobel P (2016) Accounting for protein subcellular localization: a compartmental map of the rat liver proteome. Molecular and Cellular Proteomics 16, 194-212. doi:10.1074/mcp.M116.064527  PMCID: PMC5294208

Tannous A, Boonen M, Zheng H, Zhao C, Germain C, Moore D, Sleat D, Jadot M, Lobel P. (2020) Comparative Analysis of Quantitative Mass Spectrometric Methods for Subcellular Proteomics. Journal of Proteome Research. Journal of Proteome Research 19, 1718-1730.
doi.org/10.1021/acs.jproteome.9b00862


