---
title: "Assignment of Proteins to Subcellular Locations Using protlocassign"
author: "Dirk Moore"
date: "`r Sys.Date()`"
#output: rmarkdown::html_vignette
#output: word_document

output: 
  word_document:
    reference_docx: word-styles-reference-03.docx
    
vignette: >
  %\VignetteIndexEntry{protlocassign}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```
Determining the locations of proteins in the cell is an important but complex problem. The major biochemical approach for this involves centrifugation-based methods to partially separate different organelles and other cellular compartments and then to determine the relative distribtuion of different proteins among the centrifugation fractions. The location of proteins of interest are then inferred by comparing their distribution to those of known locations. This package takes data from centrifugation fractions for large numbers of proteins to determine their relative abundance among the fractions, and by extension to their relative abundance among subcelluar locations. 

This package implements a subcellular protein assignment procedure as "constrained proportional assignmet", or CPA (Jadot et al, 2016). The input data consists of a file with one column, geneName, of gene/protein identifiers, or of peptide identifiers, and multiple columns of the relative abundances of the proteins among the centrifugal fractions. The sum of the abundance levels is constrained to 1 but can be further transfromed to improve the quality of the fit. Another key input to the program is a curated list of reference genes, and their subcellular locations. The method first determines the profiles of abundance levels of genes in a curated reference set. Then, for each gene, it finds either a profile matching a single compartment or a linear combination of compartment profiles that match the profile of that gene. The relative weights of the linear combination in principle reflect the relative abundance of a given protein among different compartments. The CPA method, unlike other previously described methods, can thus account for proteins that have multiple residences, and estimate the relative proportion among these residences. The method can also be applied to classify individual peptides, gaining potential insight into subcellular location of protein isoforms. There is also a tool to compute nearest neighbor distances of proteins, allowing grouping of proteins with similar distributions.



## Tutorial on gene assignment

We illustrate with an example. Tannous et al. (2020) presented abundance levels of proteins among six fractions from a differential centrifugation experiment: N, M, L1, L2, P, and S, and three fractions from a Nycodenz density gradient separation of the differential fraction L1 (Nyc1, Nyc2, and Nyc3). Eight subcellular compartments were considered: nucleus, mitochondria, lysosomes, peroxisomes, Golgi apparatus, plasma membrane, and cytosol. The CPA method assigns each of a large number of genes to one or more of these compartments, based on profiles from a set of reference genes. 

## Installation
Start by installing the devtools package from CRAN, by typing

```
install.packages("devtools")
```

Then install the protlocassign package from the github repository by typing

```
devtools::install_github("mooredf22/protlocassign0p1p1")
```

This will make the programs and datasets available.
Also, the `BB` library is required; it may be downloaded from CRAN by typing

```
install.packages("BB")
```
Once you have installed a package, you do not need to re-install it the next time you start up R.

These are the only packages you need to run the programs that work with data that are presented as mean levels for a protein/gene. In another vignette there is a description of R functions that work with data at the spectral level; additional packages will be needed to process spectral-level data and produce gene-level data.

In this example we will work with the TMT ms2 data from Tannous et al. (2020). To get started, load the package and look at the dimensions of the data set and then the first few rows:

```{r, echo=TRUE}
library(protlocassign)
dim(geneProfileSummaryTMTms2)
options(digits=2)
geneProfileSummaryTMTms2[1:5,1:10]
```
The data set gives relative fraction levels for 7893 genes. The last two columns give the number of spectra and the number of peptides (sequences) for each gene. (The spectra were averaged using a nested random effects model described in Jadot, 2016, to produce the given average for each gene.) Here are the numbers of spectra and peptides (sequences) from columns 11 and 12, along with the gene names in column 1:
```{r, echo=TRUE}
geneProfileSummaryTMTms2[1:5, c(1, 11:12)]
```

There are 39 reference genes, which are from Jadot et al (2016); they may be accessed as follows:
```{r, echo=TRUE}
dim(refLocProteinsJadot)
refLocProteinsJadot
```


To obtain profiles for the reference genes, use the function `cpaSetup`, which averages the spectra in each reference gene. 

```{r, echo=TRUE, eval=TRUE}
matLocRuse <- cpaSetup(geneProfileSummary=geneProfileSummaryTMTms2, refLocProteins=refLocProteinsJadot, n.channels=9)
matLocRuse

```

To view them, use `referenceProfilePlot`:
```{r, echo=TRUE, eval=TRUE, fig.show='hold'}

#par(mfrow=c(4,3))
referenceProfilePlot(refLocProteins=refLocProteinsJadot, geneProfileSummary=geneProfileSummaryTMTms2,
                     matLocR=matLocRuse, n.channels=9)
```
Now we can run the CPA routine; this may take several minutes to complete. The last column is a convergence indicator, with 1 indicating successful convergence. We remove this last column to get a matrix of gene names, and columns indicating the estimated proportional assignments of each gene among the eight subcellular locations.

```{r, echo=TRUE, eval=TRUE}
assignPropsOutTemp <- proLocAll(geneProfileSummary=geneProfileSummaryTMTms2, matLocR=matLocRuse,
                            n.channels=9)
ncol.a <- ncol(assignPropsOutTemp)   # last column is a convergence indicator, which should be 1
assignPropsUse <- data.frame(assignPropsOutTemp[,-ncol.a] ) # drop this last column
head(assignPropsUse)
```


We can look at the profile of, for example, the gene "TLN1". We can first find the gene in the dataset, and also the row number of that gene:
```{r, echo=TRUE, eval=TRUE}
#protIndex("AADAC")
protIndex("TLN1")
#protPlotfun(protPlot=93, assignProbsOut=assignProbsOut)
```
This function also accepts partial matching of the first few letters of a gene. For example, we can find the indices of the genes starting with "TLN""
```{r, echo=TRUE, eval=TRUE}
#protIndex("AAD")
protIndex("TLN")
```
Next, load the original data, with spectra and peptides, and select the columns for gene name, peptide name, and the nine fractions ranging from N to Nyc:

```{r, echo=TRUE, eval=TRUE}
finalListUse <- subset(tmtMS2orig, select={c(gene:peptide, N:Nyc.3)})
```

Now we plot the results for gene number TLN1:

```{r, echo=TRUE, eval=TRUE, fig.width=7, fig.height=7}
protPlotfun(protName="TLN1",
 geneProfileSummary=geneProfileSummaryTMTms2, finalList=finalListUse, 
 n.fractions=9, Nspectra=T, matLocR=matLocRuse, assignPropsMat=assignPropsUse)
```                        

The horizontal axis represents the nine fractions, which are N, M, L1, L2, P, S, Nyc.1, Nyc.2, and Nyc.3. In each of the eight plots, the red line is the average profile of the peptides, which are represented by the blue lines; thicker blue lines indicate larger numbers of spectra for a particular peptide. The dashed yellow-black lines show the expected profile for a protein entirely resident in the respective subcellular location. In this set of plots, we see that the CPA procedure assigns a 59 percent residence proportion to plasma membrane and 35 percent residence to cytosol. Visually, we see that the observed red profile closely matches a mixture of the expected yellow-blacklines.

## Nearest genes in a data set
We may find the genes with profiles nearest to a given gene using the function "nearestGenes". Distance is computed as the Euclidean distance between profiles. To use the function, we first create a distance matrix for the genes in a list of mean profiles, such as "geneProfileSummaryTMTms2". In this dataset, the profiles are in columns 2 through 10.

```{r, echo=TRUE, eval=TRUE}
distUse <- dist(geneProfileSummaryTMTms2[,2:10], method="euclidean")
```
Then select the gene names:

```{r, echo=TRUE, eval=TRUE}
genesUse <- geneProfileSummaryTMTms2[,1]
```
Finally, provide a gene name. Here, we first look for genes in our dataset that start with "TLN". We choose the gene "AADAC", and find the 10 nearest genes.
```{r, echo=TRUE, eval=TRUE}
protIndex("TLN")
nearestGenes("TLN1", n.nearest=10,  distGenes=distUse, geneNames=genesUse)
```

We now consider the Christofrou data, which has 20 channels (from two combined experiments).

```{r, echo=TRUE, eval=TRUE}
refLocProteinsJadotMouseMod <- refLocProteinsJadotMouse[-c(2,18),] 
# remove row 2 and 18  (GPT_MOUSE and CPSM_MOUSE)
matLocRuse <- cpaSetup(geneProfileSummary=geneProfileSummaryChristoforou,
                       refLocProteins=refLocProteinsJadotMouseMod, n.channels=20)
signif(matLocRuse, digits=2)
matLocRuse
```

Next we estimate the CPA proportions for this experiment.

```{r, echo=TRUE, eval=TRUE}
assignPropsOutTemp <- proLocAll(geneProfileSummary=geneProfileSummaryChristoforou,
                                matLocR=matLocRuse, n.channels=20)
ncol.a <- ncol(assignPropsOutTemp)   # last column is a convergence indicator, which should be 1
assignPropsUse <- data.frame(assignPropsOutTemp[,-ncol.a] ) # drop this last column
head(assignPropsUse)
```

Now we can plot the CPA estimates from the Christoforou data.

```{r, echo=TRUE, eval=TRUE}
protPlotfun(protName="TLN1_MOUSE", geneProfileSummary=geneProfileSummaryChristoforou, finalList=NULL, n.fractions=20,
            Nspectra=T, matLocR=matLocRuse, assignPropsMat=assignPropsUse)
```
This data set yields estimates of 73 percent plasma membrane and 22 percent cytosol, which are comparable to the rates from Tannous et al discussed above (59 percent and 35 percent, respectively).

## References

Christoforou A, Mulvey, CM, Breckels LM, Geladaki A, Hurrell T, Hayward PC, Naake T, Gatto L, Viner R, Arias AM, and Lilley KS (2016), "A draft map of the mouse pluripotent stem cell spatial proteome" Nature Communications 7, 8992. DOI: 10.1038/ncomms9992

Jadot M, Boonen M, Thirion J, Wang N, Xing J, Zhao C, Tannous A, Qian M, Zheng H, Everett JK, Moore DF, Sleat DE, Lobel P (2016) Accounting for protein subcellular localization: a compartmental map of the rat liver proteome. Molecular and Cellular Proteomics 16, 194-212. doi:10.1074/mcp.M116.064527  PMCID: PMC5294208

Tannous A, Boonen M, Zheng H, Zhao C, Germain C, Moore D, Sleat D, Jadot M, Lobel P. (2020) Comparative Analysis of Quantitative Mass Spectrometric Methods for Subcellular Proteomics. Journal of Proteome Research. Journal of Proteome Research 19, 1718-1730.
doi.org/10.1021/acs.jproteome.9b00862


