---
title: "Vignette 1: Tutorial on assignment of Proteins to Subcellular Locations Using protlocassign"
author: "Dirk Moore"
date: "`r Sys.Date()`"
#output: rmarkdown::html_vignette
output: 
   word_document:
#  html_document:
     #mathjax:  default
     fig_caption:  true
     toc: true
     #section_numbering: true
     #css: ggsci.css


#output: 
#  word_document:
#    reference_docx: word-styles-reference-03.docx
#output: pdf_document
#fontsize: 10pt

vignette: >
  %\VignetteIndexEntry{protlocassign}
  %\VignetteEngine{knitr::rmarkdown}
#  %\VignetteEncoding{UTF-8}
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(
  #collapse = TRUE,
  comment = "",
  fig.width = 4,
  fig.height = 4,
	message = FALSE,
	warning = FALSE,
	tidy.opts = list(
		keep.blank.line = TRUE,
		width.cutoff = 150
		),
	options(width = 150),
	eval = TRUE
)
```
Determining the locations of proteins in the cell is an important but complex problem. A frequently employed approach for this involves centrifugation-based methods to partially separate different organelles and other cellular compartments and then to determine the relative distribution of different proteins among the differential fractions. The location of proteins of interest are then inferred by comparing their distribution to reference proteins of known locations. This package takes data from centrifugation fractions for large numbers of proteins to determine their relative abundance among the fractions, and by extension to their relative abundance among subcellular locations. 

This package implements a subcellular protein assignment procedure as "constrained proportional assignment", or CPA (Jadot et al, 2016). The first column of the input data consists of the field "protName". This is a unique identifier for each row, e.g., a protein name, protein group, gene name, accession number, or peptide sequence. Following this are multiple columns of the relative abundances of the proteins among the centrifugal fractions. The sum of the abundance levels is constrained to 1 but can be further transformed to improve the quality of the fit. Another key input to the program is a curated list of reference proteins with defined subcellular locations. The method first determines the profiles of abundance levels of proteins in a curated reference set. Then, for each protein, it finds either a profile matching a single compartment or a linear combination of compartment profiles that match the profile of that protein. The relative weights of the linear combination in principle reflect the relative abundance of a given protein among different compartments. The CPA method, unlike other previously described methods, can thus account for proteins that have multiple residences, and estimate the relative proportion among these residences. The method can also be applied to classify individual peptides, gaining potential insight into subcellular location of protein isoforms. 



## Tutorial on protein assignment

We illustrate with an example. Tannous et al. (2020) presented abundance levels of proteins among six fractions from a differential centrifugation experiment (N, M, L1, L2, P, and S), and three fractions from a Nycodenz density gradient separation of the differential fraction L1 (Nyc1, Nyc2, and Nyc3). Eight subcellular compartments were considered: nucleus (Nuc), mitochondria (Mito), lysosomes (Lyso), peroxisomes (Perox), endoplasmic reticulum (ER) Golgi apparatus (Golgi), plasma membrane (PM), and cytosol (Cyto). The CPA method assigns each of a large number of proteins to one or more of these compartments, based on profiles from a set of reference genes. 

## Installing the protlocassign package
Start by installing the devtools package from CRAN, by typing

```
install.packages("devtools")
```

Then install the protlocassign package from the github repository by typing

```
devtools::install_github("mooredf22/protlocassign0p1p1")
```

This will make the programs and datasets available.
Also, the `BB` library is required; it may be downloaded from CRAN by typing

```
install.packages("BB")
```
Once you have installed a package, you do not need to re-install it the next time you start up R.

These are the only packages required for data that are presented as mean levels for a protein/gene. In another vignette there is a description of R functions that work with data at the spectral level; additional packages will be needed to process spectral-level data and produce protein-level data.

## Working with data

To use this package you will need two data sets. One consists of a column of protein names followed by a series of columns of relative protein levels derived from a subcellular fractionation experiment. (Optionally there may be two additional columns representing the numbers of peptides and spectra that were used to compute mean protein abundance levels.) The other data set consists of a list of reference proteins in the first column and the name of the known subcellular compartment in the second column.

Consider for example the TMT MS2 data from Tannous et al. (2020). To get started, load the package and look at the dimensions of the data set and then the last few rows:

```{r, echo=TRUE}
library(protlocassign)
dim(protProfileSummaryTMTms2)
```

Now use the `str` function to summarize the contents of `protProfileSummaryTMTms2`. The command `options(digits=2)` is optional; it reduces the number of significant digits displayed.

```{r, echo=TRUE}
options(digits=2)
str(protProfileSummaryTMTms2)
```

Each row represents a column, and presents the variable name, type (character, number, or integer) and the first few values of that variable. The first nine columns represent the average levels of the different fractions, across spectra. The last two columns, which are optional, give the number of spectra and the number of peptides (sequences) for each protein. (While we use a nested random effects model described in Jadot, 2016 to compute the means across spectra, other methods, including taking a straight average, may be appropriate for other applications.) 

You may also examine the first few rows by typing `head(protProfileSummaryTMTms2)`, and we see that the row names of the data set contain the protein names:

```{r, echo=TRUE}
round(head(protProfileSummaryTMTms2), digits=2)
```

The second data set we need has 39 reference proteins and two columns, the first for the protein name and the second for their respective subcellular compartments. These are from Jadot et al (2016) and may be viewed as follows:

```{r, echo=TRUE}
dim(refLocProteinsJadot)
refLocProteinsJadot
```

The protlocassign package includes several other relative fraction data sets. 

It is also possible to read in data stored as external files. Suppose for example that one is working with the Windows operating system and that the data sets reside in a directory `"C:\temp\myproteindata"`. First, set the working directory to point there:

```
setwd("C:\\temp\\myproteindata")
```

Note that in R each backslash character must be doubled to be interpreted correctly (since otherwise it will be incorrectly interpreted as an escape character). Alternatively, you may use a single forward slash:

```
setwd("C:/temp/myproteindata")
```

Then read in the two data sets (profile data and reference genes), which must be in
comma-separated format (`.csv`), with the first row containing column names:

```
protProfileSummaryMyProfilesAll <- read.csv(file="protProfileSummaryMyProfiles.csv",
                                         row.names=F)
refLocProteinsMyRefs <- read.csv(file="refLocProteinsMyRefs.csv", row.names=F)
```

The form of the data required for the `protlocassign` package require that the protein names be present only as row names, not as a column. So we create the data we need by dropping the first column (the protein names) and setting those as row names:

```
protProfileSummaryMyProfiles <- protProfileSummaryMyProfilesAll[,-1]
rownames(protProfileSummayrMyProfiles) <- protProfileSummaryMyProfilesAll[,1]
```

If one is working with either the linux or Mac OS, appropriate changes to these procedures will be needed to access files.

## Obtaining constrained proportionate assignments of proteins to compartments

In order to assign proteins proportionately to their respective compartments, we first use the function `cpaSetup` to obtain profiles for the compartments based on the mean of their reference proteins. This function produces a matrix that has one row for each compartment and one column for each fraction.

```{r, echo=TRUE, eval=TRUE}
markerLocRuse <- cpaSetup(protProfileSummary=protProfileSummaryTMTms2, refLocProteins=refLocProteinsJadot, n.channels=9)
markerLocRuse

```

To view a particular compartment profile and its component genes, use `refProfilePlot`. To examine the available compartments, use the row names of markerLocR. 

```{r, echo=TRUE, eval=TRUE, fig.show='hold'}
rownames(markerLocRuse)
```

For example, to plot the profile for plasma membrane, use this:

```{r, echo=TRUE, eval=TRUE, fig.show='hold'}

refProfilePlot(refLoc="PM", refLocProteins=refLocProteinsJadot,
                     protProfileSummary=protProfileSummaryTMTms2,
                     markerLocR=markerLocRuse)
```

We see, in the plot title, that there are five reference proteins. To plot, for example, the second one, enter this code, using the option `refProtPlot=2`:

```{r, echo=TRUE, eval=TRUE, fig.show='hold'}

refProfilePlot(refLoc="PM", refLocProteins=refLocProteinsJadot,
                     protProfileSummary=protProfileSummaryTMTms2,
                     markerLocR=markerLocRuse, refProtPlot=2)
```

To show all five of the reference proteins, open a window 10 inches by 7 inches. Then use `par(mfrow=c(3,2))` to set up a plot array with 3 rows and 2 columns, to allow plotting all five reference proteins on a single page. Finally, plot all five of the them one-by-one by specifying `for (j in 1:5)`. (If using plain R rather than R Studio, to  to fit them all on the same page, you may need to open a window by entering `windows(width=7, height=10)`.) The parameters in `par(mfrow=c(3,2))` and `for (j in 1:5)` can be adjusted to display any number of plots.


```{r, echo=TRUE, eval=TRUE, fig.show='hold', fig.width = 7, fig.height = 10}

par(mfrow=c(3,2))
for (j in 1:5) {
  refProfilePlot(refLoc="PM", refLocProteins=refLocProteinsJadot,
                     protProfileSummary=protProfileSummaryTMTms2,
                     markerLocR=markerLocRuse, refProtPlot=j)
  }
```

To plot all of the reference profiles we may set up a plot window with four rows and two columns and then loop through the eight subcellular compartments.

```{r, echo=TRUE, eval=TRUE, fig.width = 7, fig.height = 10}
loc.list <- rownames(markerLocRuse)
n.loc <- length(loc.list)
par(mfrow=c(4,2))
for (i in 1:n.loc) {
  refProfilePlot(refLoc=loc.list[i], refLocProteins=refLocProteinsJadot,
                     protProfileSummary=protProfileSummaryTMTms2,
                     markerLocR=markerLocRuse)
  }
```

Now we can run the CPA routine, which assigns proteins to their respective compartments; this may take several minutes to complete. 

```{r, echo=TRUE, eval=TRUE, message=FALSE, warning=FALSE}
assignPropsUse <- proLocAll(protProfileSummary=protProfileSummaryTMTms2,
                            markerLocR=markerLocRuse, n.channels=9)

str(assignPropsUse)
```

The first column contains the protein names, and the next 8 columns contain the allocation proportions of each protein to the eight compartments. The last two columns are integers representing the numbers of peptides and spectra, which are carried over from the input `protProfileSummary` profile data, these are not essential for the CPA analysis.

We can look at the profile of, for example, the protein "TLN1". We first ensure that the protein is in the dataset:

```{r, echo=TRUE, eval=TRUE}
protIndex("TLN1", protProfileSummary=protProfileSummaryTMTms2)
```

This function also accepts partial matching of the first few letters of a protein. For example, we can find the indices of the genes starting with "TLN":

```{r, echo=TRUE, eval=TRUE}
protIndex("TLN")
```



Now we plot the results for protein TLN1:

```{r, echo=TRUE, eval=TRUE, fig.width=7, fig.height=9}
protPlotfun(protName="TLN1", protProfileSummary=protProfileSummaryTMTms2, 
n.fractions=9, n.compartments=8,
 Nspectra=T, markerLocR=markerLocRuse, assignPropsMat=assignPropsUse)
```                        

The horizontal axis represents the nine fractions, which are N, M, L1, L2, P, S, Nyc.1, Nyc.2, and Nyc.3. In each of the eight plots, the red line is the average profile of the protein. The dashed yellow-black lines show the expected profile for a protein entirely resident in the respective subcellular location. In this set of plots, we see that the CPA procedure assigns a 59 percent residence proportion to plasma membrane and 35 percent residence to cytosol. Visually, we see that the observed red profile closely matches a mixture of the expected yellow-blacklines. 

The `protPlotfun` function is designed to plot profiles of eight subcellular locations. If a data set has more than eight of these, it will be necessary to modify the code to accommodate the larger number.

## Saving the CPA output
### Data output

We may save the results of the `proLocAll` procedure by first setting a directory.

```
setwd("C:/temp/myProteinOutput")
```
The write out the data as a comma separated file, using `write.csv`:

```
write.csv(assignPropsUse, file="assignPropsUse.txt", row.names=F)
```

This writes the CPA results (which we had previously put into the R dataset `assignPropsUse`), without row names, and in comma separated format. Ordinarily one would give the file a `.csv` extension so that Excel would automatically open it when clicking on the file name in Windows. Here we do not recommend that because of a naming issue with proteins. Specifically, some proteins (`MARCH9` for example) will be interpreted by Excel as dates, and switched to a date format (`9-Mar` in this case). To prevent that, save the comma-separated value file as a text file, with extension `.txt`. Then, in Windows, open up the Excel program and navigate within Excel to the `assignPropsUse.txt` file and open it. You will then be presented with an Excel import wizard where you specify that columns are separated by commas, and then select the first row (`protName`), and specify that this is a text column. This will prevent Excel from converting protein names to date format.

### Plot output as pdf files

To save the plot of a protein (`TLN1` for example) as a pdf file, we first specify a pdf plot `window`, call the `protPlotfun` function as before, and then close the plotwindow using `dev.off()` to allow R to complete producing the file:

```
pdf(file="myPlotPDFfile.pdf", width=7, height=10)
protPlotfun(protName="TLN1", protProfileSummary=protProfileSummaryTMTms2, 
    n.fractions=9, n.compartments=8,
    Nspectra=T, markerLocR=markerLocRuse, assignPropsMat=assignPropsUse)
dev.off()
```   

To output plots all of the protein profiles into a single pdf file, one can set up a loop as follows:


```
pdf(file="allPlotsPDFfile.pdf", width=7, height=10)
n.prots <- nrow(assignPropsUse)
for (i in 1:n.prots) {
   protPlotfun(protName=rownames(assignPropsUse)[i],
       protProfileSummary=protProfileSummaryTMTms2, n.fractions=9, n.compartments=8,
       Nspectra=T, markerLocR=markerLocRuse, assignPropsMat=assignPropsUse)
}
dev.off()
```   

This will result in a single file, `allPlotsPDFfile.pdf` with 7893 pages, one for each protein.




## References

Jadot M, Boonen M, Thirion J, Wang N, Xing J, Zhao C, Tannous A, Qian M, Zheng H, Everett JK, Moore DF, Sleat DE, Lobel P (2016) Accounting for protein subcellular localization: a compartmental map of the rat liver proteome. Molecular and Cellular Proteomics 16, 194-212. doi:10.1074/mcp.M116.064527  PMCID: PMC5294208

Tannous A, Boonen M, Zheng H, Zhao C, Germain C, Moore D, Sleat D, Jadot M, Lobel P. (2020) Comparative Analysis of Quantitative Mass Spectrometric Methods for Subcellular Proteomics. Journal of Proteome Research. Journal of Proteome Research 19, 1718-1730.
doi.org/10.1021/acs.jproteome.9b00862


